Firebase API

kusera opened the pull request.

Commit:
Add ignore for local firebase json

Commit:
Add Firebase dependency

Commit:
Add skeleton for ChatAPI

Commit:
Change 'close' to 'delete'

Commit:
Add server action handler

- Changed VIEW-ACTION string to VIEW_ACTION

Commit:
Add API to handle firebase server interaction

Commit:
Add constant for room fetches

Commit:
Add action to dispatch room fetches

Commit:
Extract room conversion function from rendering

Commit:
Remove extraneous data attributes

Commit:
Add room fetch from server at app start

Commit:
Register response to room fetch

Commit:
Add constant for room submission

Commit:
Remove outdated properties for room creation

Commit:
Remove RoomActions require

Commit:
Cleaned up code organization

- Composite paths into single object
- Made error message slightly more descriptive

chikamichi commented:
By the way, I guess you could rename the PR (which borrowed the name from first commit).

kusera commented:
> PR (which borrowed the name from first commit).
@chikamichi What does PR refer to?

Commit:
Add skeleton for creating rooms on the firebase

- Made the language in RoomActions.js more consistent

Commit:
Change name of convertMessage()

chikamichi commented:
*Pull Request*.

Commit:
Moved RoomNode creation via object to RoomeNode

Commit:
Remove extra whitespace

Commit:
Listen to firebase for changes during first fetch

- Bug: ChatAPI fires off a ROOM_FETCHED dispatch before NEW_ROOM_SUBMIT
  finishes, throwing an error.
- Bug: RoomStore's setRoom() should be changed to replace its current list
  rather than appending to it (creates duplicate entries on subsequent
  fetches)

kusera commented:
Option 1: setTimeOut()
I don't really like this option because it only masks the issue, it doesn't solve it. The connection could, for example, hiccup, and then delay the server listener long enough to exceed the timeout, which should be pretty short for UX.

Option 2: Action Queue
Instead of dispatching directly my action creators could send them into a queue and then the Dispatcher runs down through them one by one. One thread I saw said this isn't ideal because it promotes bad coding practices (nesting dispatches deeply).

Option 3: Two dispatchers?
Unorthodox, yes, but if I have one Dispatcher for view actions and another for server actions it might help prevent conflicts. Also I have no idea if it'd work.

Commit:
Unify dispatch handlers and payload names

chikamichi commented:
@kusera I don't trigger an actual error with `ROOMS_FETCHED` vs `NEW_ROOM_SUBMIT`. What exactly is the problem here?

chikamichi commented:
Do you mean that, if the user tries to create a new room while the list of rooms is being fetched, and that list arrives *before* the new room is persisted, the list is truncated and the UI incorrectly does not report the newly created room?

kusera commented:
On app start, ChatAPI will fetch a list of rooms. This works fine.

When I click the create room button, however, it will hear a `value` event from the server and dispatch a `ROOM_FETCHED` while the `NEW_ROOM_SUBMIT` dispatch is still going on and this causes an error on my side.

kusera commented:
Something like:

Click > new room submit action triggers > api sends a room to firebase > firebase  creates a room on the server > firebase notifies api of update > api fires a room fetch action while the new room submit dispatch is live/active

kusera commented:
The error I get is "invariant Violation: Cannot dispatch in the middle of a dispatch." 

I know from logging the events that it's because the API tries to fire ROOMS_FETCH right after NEW_ROOM_SUBMIT

chikamichi commented:
Hum, I don't get that error. If it's still lurking later on today, you'll have to show me.

chikamichi commented:
That kind of error usually happens when an action A is currently being dispatched, and as a side-effect, an action B is required to be dispatched before A has been fully dispatched. Honestly, I don't really know what "fully dispatched" means from the dispatcher's perspective. I only know the dispatcher we're borrowing from flux is synchronous, in that only one action may be dispatch at a given time. For I don't really know the synchronicity criterium, I'm not sure what's wrong with your code.

I know how to workaround it, though. You may trick the dispatcher into thinking two nested actions are actually independent, by wrapping the nested one (B) in its own execution context (a JS closure). One may do that by using `setTimeout(actionB, 0)`. If you want to dig deeper, [here's something of interest](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/), but be aware it's quite advanced and refers to the concept of defrerable/promises which you're not familiar with I believe. Closed parenthesis.

Commit:
Fix RoomStore not replacing its _rooms

Commit:
Remove outdated functions and dispatch registers

Commit:
Change 'submitRoom' to 'createRoom' everywhere

Commit:
Change _room clear to superior hack

Commit:
Listen to servers only once for getRoom

- Change 'err' to 'error' for future consistency

Commit:
Prevent room rewrites if it already exists

Commit:
Solve conflicting dispatch issue (for now)

- Split current and future server actions to a different Action Creator
- ChatAPI now "intercepts" certain actions from RoomActions, rather than
  listening for them by registering with Dispatcher
- Add placeholder function for room deletion pending component support
